integration/run_test.go

```diff
@@ -185,3 +191,216 @@ func testWebServiceContainer(shell *shell.Shell, containerName string) {
+func TestNetworkRetry(t *testing.T) {
+
+	const containerImage = alpineImage
+
+	tests := []struct {
+		name   string
+		config retryConfig
+	}{
+		{
+			name: "No network interruption, no retries allowed, success",
+			config: retryConfig{
+				MaxRetries:         -1,
+				MinWaitMsec:        0,
+				MaxWaitMsec:        0,
+				NetworkDisableMsec: 0,
+				ExpectedSuccess:    true,
+			},
+		},
+		{
+			name: "1s network interruption, no retries allowed, failure",
+			config: retryConfig{
+				MaxRetries:         -1,
+				MinWaitMsec:        0,
+				MaxWaitMsec:        0,
+				NetworkDisableMsec: 1000,
+				ExpectedSuccess:    false,
+			},
+		},
+		{
+			name: "2s network interruption, ~3-4s retries allowed, success",
+			config: retryConfig{
+				MaxRetries:         5,
+				MinWaitMsec:        100,
+				MaxWaitMsec:        1600,
+				NetworkDisableMsec: 2000,
+				ExpectedSuccess:    true,
+			},
+		},
+		{
+			name: "5s network interruption, ~1-2s retries allowed, failure",
+			config: retryConfig{
+				MaxRetries:         1,
+				MinWaitMsec:        100,
+				MaxWaitMsec:        1600,
+				NetworkDisableMsec: 5000,
+				ExpectedSuccess:    false,
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			regConfig := newRegistryConfig()
+			sh, done := newShellWithRegistry(t, regConfig)
+			defer done()
+
+			registryHostIP := getIP(t, sh, regConfig.host)
+
+			config := `
+[blob]
+max_retries   = ` + strconv.Itoa(tt.config.MaxRetries) + `
+min_wait_msec = ` + strconv.FormatInt(tt.config.MinWaitMsec, 10) + `
+max_wait_msec = ` + strconv.FormatInt(tt.config.MaxWaitMsec, 10) + `
+
+[background_fetch]
+disable = true
+`
+
+			m := rebootContainerd(t, sh, getContainerdConfigToml(t, false), getSnapshotterConfigToml(t, false, config))
+			// Mirror image
+			copyImage(sh, dockerhub(containerImage), regConfig.mirror(containerImage))
+			// Pull image, create SOCI index with all layers and small (100kiB) spans
+			indexDigest := buildIndex(sh, regConfig.mirror(containerImage), withMinLayerSize(0), withSpanSize(100*1024))
+			sh.X("soci", "image", "rpull", "--user", regConfig.creds(), "--soci-index-digest", indexDigest, regConfig.mirror(containerImage).ref)
+
+			// Run the container
+			image := regConfig.mirror(containerImage).ref
+			sh.X("soci", "run", "-d", "--snapshotter=soci", image, "test-container")
+
+			sh.X("apt-get", "--no-install-recommends", "install", "-y", "iptables")
+
+			// TODO: Wait for the container to be up and running
+
+			// Block network access to the registry
+			if tt.config.NetworkDisableMsec > 0 {
+				sh.X("iptables", "-A", "OUTPUT", "-d", registryHostIP, "-j", "DROP")
+			}
+
+			// Do something in the container that should work without network access
+			commandSucceedStdout, _, err := sh.R("ctr", "task", "exec", "--exec-id", "test-task-1", "test-container", "sh", "-c", "times")
+			if err != nil {
+				t.Fatalf("attempt to run task without network access failed: %s", err)
+			}
+
```

Comment:
nit: looks like this is only used once; maybe have a anon struct

Reply:
This weighs on one of my refactor ideas. If we're going to have multiple different log monitor functions doing json decoding, I think we should switch from explicit struct decoding to more generic map decoding. That would allow a single function within the log monitor to handle arbitrary "when key X has value Y do thing Z" conditions, which seem like they will be the majority of cases.

Reply:
Resolving in favor of parallel discussion thread.